src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -888,8 +1024,17 @@ ModifyQuerySupported(Query *queryTree, Query *originalQuery, bool multiShardQuer
 					 PlannerRestrictionContext *plannerRestrictionContext)
 {
 	Oid distributedTableId = InvalidOid;
-	DeferredErrorMessage *error = ModifyPartialQuerySupported(queryTree, multiShardQuery,
-															  &distributedTableId);
+	DeferredErrorMessage *error = MergeQuerySupported(originalQuery,
+													  plannerRestrictionContext);
+	if (error)
+	{
+		/*
```

Comment:
hmmm, not supporting recursive planning makes this patch a little simpler, but also restricts lots of useful query patterns. Also, creates a non-unified user experience (e.g., why CTEs do not work with merge command?)

What are the problems of allowing recursive planning? I don't see many. One potential problem I can see is that some parts of the recursive planning operates on `query->rtable`, which includes the target table. But I believe those parts should also be similar for UPDATE/DELETE queries -- which is fine from recursive planner.

Majority of the recursive planner operates on `query->jointree`, which only includes the "source" part for the merge command. That could help supporting arbitrarily complex queries in the "source", which could be useful.

(note: I'm fine to omit recursive planning in this patch, but seems like we should consider as a follow-up PR for phase-3).

Reply:
I was conservative about this check, but we do support CTEs as source with this patch (unit tests are added in this PR). I want to understand a bit more about the recursive planning scenarios, let's talk about it and we can remove this.

Reply:
Removed it.